File: services/rag-engine/src/reasoning/followup_detector.py
class FollowupDetector:
    """Heuristic follow-up detection. No LLM call needed."""
    
    PRONOUN_PATTERNS = [
        "it", "this", "that", "them", "these", "those",
        "its", "the same", "above", "previous", "earlier",
        "in java", "in python", "in c++", "in javascript", "in go", "in rust",
    ]
    
    def is_followup(self, query: str, history: list) -> bool:
        if not history:
            return False
        
        q = query.lower().strip()
        
        # Short queries with history are likely follow-ups
        if len(q.split()) <= 5 and history:
            return True
        
        # Contains pronouns that reference previous context
        for pattern in self.PRONOUN_PATTERNS:
            # Match whole words or boundary checks
            import re
            if re.search(r'\b' + re.escape(pattern) + r'\b', q):
                return True
        
        # Starts with conjunctions suggesting continuation
        if q.startswith(("and ", "but ", "also ", "what about ", "how about ", "why ", "how to ")):
            # "how to" or "why" as short sentences are often follow-ups if history is present
            if len(q.split()) <= 6:
                return True
        
        return False

File: services/rag-engine/src/main/java/com/llmops/rag/service/FollowupDetector.java
package com.llmops.rag.service;

import com.llmops.rag.model.ChatMessage;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.regex.Pattern;

@Service
public class FollowupDetector {

    private static final List<String> PRONOUN_PATTERNS = List.of(
            "it", "this", "that", "them", "these", "those",
            "its", "the same", "above", "previous", "earlier",
            "in java", "in python", "in c++", "in javascript", "in go", "in rust"
    );

    public boolean isFollowup(String query, List<ChatMessage> history) {
        if (history == null || history.isEmpty()) {
            return false;
        }

        String q = query.toLowerCase().trim();

        // Short queries with history are likely follow-ups
        String[] words = q.split("\\s+");
        if (words.length <= 5) {
            return true;
        }

        // Contains pronouns referencing previous context
        for (String pronoun : PRONOUN_PATTERNS) {
            Pattern pattern = Pattern.compile("\\b" + Pattern.quote(pronoun) + "\\b");
            if (pattern.matcher(q).find()) {
                return true;
            }
        }

        // Starts with conjunctions suggesting continuation
        if (q.startsWith("and ") || q.startsWith("but ") || q.startsWith("also ") || 
            q.startsWith("what about ") || q.startsWith("how about ") || 
            q.startsWith("why ") || q.startsWith("how to ")) {
            if (words.length <= 6) {
                return true;
            }
        }

        return false;
    }
}
